File: python/dmp_base/FunctionApproximator.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 30 19:00:00 2017

@author: gsutanto
"""

import numpy as np
import os
import sys
import copy
from CanonicalSystem import *

class FunctionApproximator:
    'Base class for function approximators of DMPs.'
    
    def __init__(self, dmp_num_dimensions_init, model_size_init, canonical_system, name=""):
        self.name = name
        self.dmp_num_dimensions = dmp_num_dimensions_init
        self.model_size = model_size_init
        self.canonical_sys = canonical_system
        self.weights = np.zeros((self.dmp_num_dimensions, self.model_size))
    
    def isValid(self):
        assert (self.dmp_num_dimensions > 0), "self.dmp_num_dimensions=" + str(self.dmp_num_dimensions) + "<= 0 (invalid!)"
        assert (self.model_size > 0), "self.model_size=" + str(self.model_size) + "<= 0 (invalid!)"
        assert (self.canonical_sys != None), "CanonicalSystem canonical_sys does NOT exist!"
        assert (self.canonical_sys.isValid() == True), "CanonicalSystem canonical_sys is invalid!"
        assert ((self.weights.shape[0] == self.dmp_num_dimensions) and (self.weights.shape[1] == self.model_size)), "Weights matrix dimensions=" + str(self.weights.shape[0]) + "X" + self.weights.shape[1] + " is/are mis-matched with self.dmp_num_dimensions=" + str(self.dmp_num_dimensions) + " and/or self.model_size=" + str(self.model_size)
        return True

File: src/dmp_base/FunctionApproximator.cpp
#include "dmp/dmp_base/FunctionApproximator.h"

namespace dmp {

FunctionApproximator::FunctionApproximator()
    : dmp_num_dimensions(0),
      model_size(0),
      canonical_sys(NULL),
      weights(new MatrixNxM(MAX_DMP_NUM_DIMENSIONS, MAX_MODEL_SIZE)),
      rt_assertor(NULL) {}

FunctionApproximator::FunctionApproximator(uint dmp_num_dimensions_init,
                                           uint model_size_init,
                                           CanonicalSystem* canonical_system,
                                           RealTimeAssertor* real_time_assertor)
    : dmp_num_dimensions(dmp_num_dimensions_init),
      model_size(model_size_init),
      canonical_sys(canonical_system),
      weights(new MatrixNxM(dmp_num_dimensions, model_size)),
      rt_assertor(real_time_assertor) {}

bool FunctionApproximator::isValid() {
  if (rt_assert((rt_assert(dmp_num_dimensions > 0)) &&
                (rt_assert(dmp_num_dimensions <= MAX_DMP_NUM_DIMENSIONS))) ==
      false) {
    return false;
  }
  if (rt_assert((rt_assert(model_size > 0)) &&
                (rt_assert(model_size <= MAX_MODEL_SIZE))) == false) {
    return false;
  }
  if (rt_assert(rt_assert(canonical_sys != NULL) &&
                rt_assert(weights != NULL)) == false) {
    return false;
  }
  if (rt_assert(canonical_sys->isValid()) == false) {
    return false;
  }
  if (rt_assert((rt_assert(weights->rows() == dmp_num_dimensions)) &&
                (rt_assert(weights->cols() == model_size))) == false) {
    return false;
  }
  return true;
}

uint FunctionApproximator::getDMPNumDimensions() { return dmp_num_dimensions; }

uint FunctionApproximator::getModelSize() { return model_size; }

CanonicalSystem* FunctionApproximator::getCanonicalSystemPointer() {
  return canonical_sys;
}

FunctionApproximator::~FunctionApproximator() {}

}  // namespace dmp
